File: src/main/scala/ru/tinkoff/gatling/influxdb/Annotations.scala
package ru.tinkoff.gatling.influxdb

import io.gatling.core.scenario.Simulation
import io.gatling.core.Predef._
import io.gatling.core.structure.{ChainBuilder, PopulationBuilder, ScenarioBuilder}
import io.razem.influxdbclient.Point
import io.gatling.core.session.el._

/**
  * Mix this trait in Simulation class to write Start/Stop annotations in influxDb before/after simulation run
  */
trait Annotations {
  simulation: Simulation =>
  {
    simulation.before(AnnotationManager.addStatusAnnotation(Start))

    simulation.after(AnnotationManager.addStatusAnnotation(Stop))
  }
}

/**
  * Use this object to write custom points to InfluxDB
  */
object Annotations {

  def influxDataPoint(tagKey: String, tagValue: String, fieldKey: String, fieldValue: String) = {
    AnnotationManager.addCustomAnnotation(tagKey, tagValue, fieldKey, fieldValue)
  }

  def influxDataPoint(point: Point) = {
    AnnotationManager.addCustomPoint(point)
  }

  def influxDataPoint(point: Seq[Point]) = {
    AnnotationManager.addCustomPoints(point)
  }

  //for write custom points from setUp
  def userDataPoint(uniqScnName: String, point: Point): PopulationBuilder = {
    //TODO: how to do it without scenario?
    scenario(uniqScnName)
      .userDataPoint(point)
      .inject(atOnceUsers(1))
  }

  //for write default prepared points from setUp
  def userDataPoint(uniqScnName: String,
                    tagKey: String,
                    tagValue: String,
                    fieldKey: String,
                    fieldValue: String): PopulationBuilder = {
    //TODO: how to do it without scenario?
    scenario(uniqScnName)
      .userDataPoint(tagKey, tagValue, fieldKey, fieldValue)
      .inject(atOnceUsers(1))
  }

  //for usage in chain builder
  implicit class ChainAppender(cb: ChainBuilder) {
    def userDataPoint(point: Point): ChainBuilder =
      cb.exec(session => {
        influxDataPoint(point)
        session
      })

    def userDataPoint(tagKey: String, tagValue: String, fieldKey: String, fieldValue: String): ChainBuilder =
      cb.exec(session => {
        for {
          tag   <- tagValue.el[String].apply(session)
          field <- fieldValue.el[String].apply(session)
        } yield influxDataPoint(tagKey, tag, fieldKey, field)
        session
      })
  }

  //for usage in scenario
  implicit class ScenarioAppender(sb: ScenarioBuilder) {
    def userDataPoint(point: Point): ScenarioBuilder =
      sb.exec(session => {
        influxDataPoint(point)
        session
      })

    def userDataPoint(tagKey: String, tagValue: String, fieldKey: String, fieldValue: String): ScenarioBuilder =
      sb.exec(session => {
        for {
          tag   <- tagValue.el[String].apply(session)
          field <- fieldValue.el[String].apply(session)
        } yield influxDataPoint(tagKey, tag, fieldKey, field)
        session
      })
  }

  //for usage in simulation setUp
  implicit class PopulationBuilderAppender(pb: PopulationBuilder) {
    def userDataPoint(uniqScpName: String, point: Point): PopulationBuilder = {
      pb.andThen(Annotations.userDataPoint(uniqScpName, point))
    }

    def userDataPoint(uniqScpName: String,
                      tagKey: String,
                      tagValue: String,
                      fieldKey: String,
                      fieldValue: String): PopulationBuilder = {
      pb.andThen(Annotations.userDataPoint(uniqScpName, tagKey, tagValue, fieldKey, fieldValue))
    }
  }

}

File: src/main/java/ru/tinkoff/gatling/javaapi/influxdb/Annotations.java
package ru.tinkoff.gatling.javaapi.influxdb;

import io.gatling.javaapi.core.*;
import io.razem.influxdbclient.*;

import java.math.BigDecimal;
import java.util.function.Function;
import java.util.ArrayList;

import static scala.jdk.CollectionConverters.IterableHasAsScala;
import static io.gatling.javaapi.core.internal.Expressions.*;


public final class Annotations {
    public static Tag Tag(String key, String value) {
        return Tag.apply(key, value);
    }

    public static Field Field(String key, String value) {
        return StringField$.MODULE$.apply(key, value);
    }

    public static Field Field(String key, double value) {
        return DoubleField$.MODULE$.apply(key, value);
    }

    public static Field Field(String key, long value) {
        return LongField$.MODULE$.apply(key, value);
    }

    public static Field Field(String key, boolean value) {
        return BooleanField$.MODULE$.apply(key, value);
    }

    public static Field Field(String key, BigDecimal value) {
        return BigDecimalField$.MODULE$.apply(key, new scala.math.BigDecimal(value));
    }

    public static Point Point(String key) {
        return new Point(
                key,
                System.currentTimeMillis() * 1000000,
                IterableHasAsScala(new ArrayList<Tag>()).asScala().toSeq(),
                IterableHasAsScala(new ArrayList<Field>()).asScala().toSeq()
        );
    }

    public static Point Point(String key, long timestamp) {
        return new Point(
                key,
                timestamp,
                IterableHasAsScala(new ArrayList<Tag>()).asScala().toSeq(),
                IterableHasAsScala(new ArrayList<Field>()).asScala().toSeq()
        );
    }

    public static Point Point(String key, long timestamp, Iterable<Tag> tags, Iterable<Field> fields) {
        return new io.razem.influxdbclient.Point(
                key,
                timestamp,
                IterableHasAsScala(tags).asScala().toSeq(),
                IterableHasAsScala(fields).asScala().toSeq()
        );
    }

    public static PopulationBuilder userDataPoint(String uniqScpName, Point point) {
        return new PopulationBuilder(ru.tinkoff.gatling.influxdb.Annotations.userDataPoint(uniqScpName, point));
    }

    public static PopulationBuilder userDataPoint(String uniqScpName,
                                                  String tagKey,
                                                  String tagValue,
                                                  String fieldKey,
                                                  String fieldValue) {
        return new PopulationBuilder(
                ru.tinkoff.gatling.influxdb.Annotations.userDataPoint(
                        uniqScpName,
                        tagKey,
                        tagValue,
                        fieldKey,
                        fieldValue)
        );
    }

    public static Function<Session, Session> userDataPoint(Point point) {
        return (Session session) -> {
            ru.tinkoff.gatling.influxdb.Annotations.influxDataPoint(point);
            return session;
        };
    }

    public static Function<Session, Session> userDataPoint(String tagKey,
                                                           String tagValue,
                                                           String fieldKey,
                                                           String fieldValue) {
        return (Session session) -> {
            toStringExpression(tagValue).apply(session.asScala()).onSuccess((tag) -> {
                toStringExpression(fieldValue).apply(session.asScala()).onSuccess((field) ->
                {
                    ru.tinkoff.gatling.influxdb.Annotations.influxDataPoint(tagKey, tag, fieldKey, field);
                    return 0;
                });
                return 0;
            });
            return session;
        };
    }

}
